### PR TITLE
[ENG-859] Auction

### PR DESCRIPTION
**Pull request by maximgeerinck**
# Description

Add Auction capabilities:

- Creating an auction (1155 and 721)
- Ending existing auctions
- Cancelling auctions

### CODE REVIEW

File: public/listForAuction.csv [hunk 1]
```diff
@@ -0,0 +1,4 @@
+ContractAddress,TokenId,Type,Currency,Price,Supply
```
Comment (mihaiberq) [inline public/listForAuction.csv:1]: what is this used for? can't seem to find its reference in the code. if it's still used, should we store it as json maybe so we dont parse it every time?

**Review by mihaiberq — APPROVED** [review]
one small question, but otherwise this looks good

File: README.md [hunk 2]
```diff
@@ -7,7 +7,7 @@
 ## TL;DR - Start on Testnet
 
 1. `yarn install`
-2. `NODE_ENV npx ts-node src/examples/mint/mint721.ts`
+2. `NODE_ENV npx esno src/examples/mint/mint721.ts`
```
Comment (Superjo149) [inline README.md:10]: Whats the difference?

File: README.md [hunk 3]
```diff
@@ -110,6 +110,54 @@ const nft = await refinable.createNft(TOKEN_TYPE.ERC721, {
...
+### Ending an auction
+
+Get the auction id and then end it.
+
+[Example](./src/examples/cancelAuction.ts)
+
+```javascript
+const auctionId = await nft.getAuctionId();
+await nft.endAuction(auctionId);
+```
```
Comment (Superjo149) [inline README.md:159]: This might be a bit confusing. I think it would also be best if we explained the differences

File: src/contracts.ts [hunk 4]
```diff
@@ -111,8 +37,8 @@ export function getERC20Contract(currency: REFINABLE_CURRENCY) {
   ]);
 }
 
-export function getERC20Address(currency: REFINABLE_CURRENCY) {
-  const address = tokenAddresses["mainnet"]?.[currency];
+export function getERC20Address(chainId: number, currency: PriceCurrency) {
```
Comment (Superjo149) [inline src/contracts.ts:40]: A lot of this functionality in this file is actually more up-to-date in `@refinableco/lib`. So in the future, ideally we would also publish or share this package in some way I think.

File: src/nft/AbstractNFT.ts [hunk 5]
```diff
@@ -133,8 +161,8 @@ export abstract class AbstractNFT {
     price: Price,
     spenderAddress: string
   ): Promise<any> {
-    if (price.currency !== REFINABLE_CURRENCY.BNB) {
-      const erc20Contract = getERC20Contract(price.currency);
+    if (price.currency !== PriceCurrency.Bnb) {
```
Comment (Superjo149) [inline src/nft/AbstractNFT.ts:164]: This is no longer correct as this logic does not work multi-chain. In our code-base we use `isNativeCurrency` for this.

File: src/nft/AbstractNFT.ts [hunk 6]
```diff
@@ -153,17 +181,79 @@ export abstract class AbstractNFT {
...
+    const startPrice = ethers.utils
+      .parseEther(price.amount.toString())
+      .toString();
+
+    const paymentToken = this.getPaymentToken(price.currency);
+
+    const blockchainAuctionResponse = await this.auctionContract.createAuction(
+      this.item.contractAddress,
+      // TODO: Preparation for V2
+      // ethers.constants.AddressZero, // _royaltyToken
```
Comment (Superjo149) [inline src/nft/AbstractNFT.ts:214]: Yeah, this is a new addition. We'll soon run into versioning issues I think.

**Review by Superjo149 — CHANGES_REQUESTED** [review]
Couple of remarks

File: README.md (hunk 2 above)
Comment (maximgeerinck) [reply README.md:10]: it is normally up to 5 times faster, but it doesn't always work, so i'll revert it :)

File: README.md (hunk 3 above)
Comment (maximgeerinck) [reply README.md:159]: you're right, adjusting

File: src/contracts.ts (hunk 4 above)
Comment (maximgeerinck) [reply src/contracts.ts:40]: That's an internal package right now :) Let's see how we can merge this when we integrate it in our codebase

File: src/nft/AbstractNFT.ts (hunk 5 above)
Comment (maximgeerinck) [reply src/nft/AbstractNFT.ts:164]: Will rework for multi chain according to new changes

File: public/listForAuction.csv (hunk 1 above)
Comment (maximgeerinck) [reply public/listForAuction.csv]: Was old, not used anymore. deleted it :)

**Review by deveshrawat — APPROVED** [review]
Looks ok to me!

**Review by Superjo149 — APPROVED** [review]

**Pull request merged**